### PR TITLE
Can now Edit and Delete a Multiple Choice Card/Deck

### PR DESCRIPTION
**Pull request by rekandiah**
#136

### CODE REVIEW

**Comment by j-suguma** [comment]
CI build passes for editDeleteMC
image

tests:
image

No static analysis since the commit for SA was not merged into master when the code for this branch was developed

**Comment by rekandiah** [comment]
I updated the branch with dev! may you try with the static analysis tool now plis

**Comment by j-suguma** [comment]
SA report for the files that were changed in this PR
[editDeleteMCSA.pdf](https://github.com/jameelbrochu/AnyMemo/files/1884593/editDeleteMCSA.pdf)

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java [hunk 1]
```diff
@@ -340,9 +340,17 @@ public void insertMultipleChoiceCard(MultipleChoiceCard mcCard, SQLiteDatabase d
     }
 
     public void deleteMultipleChoiceCard(MultipleChoiceCard mcCard) {
-        db.delete(MultipleChoiceContract.MultipleChoiceCardTable.TABLE_NAME,
-                "id" + "=" + mcCard.getId(),
-                null);
+        Cursor res = db.rawQuery("SELECT * FROM " + MultipleChoiceContract.MultipleChoiceCardTable.TABLE_NAME,null);
```
Comment (mukulikadey) [inline app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: Can you explain what the cursor does again please?

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java [hunk 2]
```diff
@@ -340,9 +340,17 @@ public void insertMultipleChoiceCard(MultipleChoiceCard mcCard, SQLiteDatabase d
...
+        int value = res.getColumnIndex("id");
+        if (value != -1) {
+            db.delete(MultipleChoiceContract.MultipleChoiceCardTable.TABLE_NAME,
```
Comment (mukulikadey) [inline app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: Why are we checking if the value is not equal to -1? Why -1 i don't get it..

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java [hunk 3]
```diff
@@ -105,38 +110,66 @@ public void onConfigurationChanged(Configuration newConfig) {
...
+
     }
 
     View.OnClickListener saveOnClickListener = new View.OnClickListener() {
 
         @Override
         public void onClick(View v) {
-            updateViews();
-            multipleChoiceCardDao.addMultipleChoiceCard(currentMCCard);
+            if(cardToUpdate == null) {
```
Comment (mukulikadey) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java]: Space after the "if"

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/OpenActionsMCFragment.java [hunk 4]
```diff
@@ -83,6 +100,35 @@ public void onClick(View v) {
                 startActivity(myIntent);
                 recentListUtil.addToRecentList(dbPath);
             }
+
+            if(v == deleteItem) {
+                new AlertDialog.Builder(mActivity)
```
Comment (mukulikadey) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/OpenActionsMCFragment.java]: Space after the "if"

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java (hunk 1 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: The cursor contains the result we obtained from the query. When using the cursor the app will have read and write access I believe lol.

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java (hunk 2 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: When a column doesn't exist then the result will be -1. So basically if the column exists I'm conducting my delete statement. I need to do this because my unit tests look for the column "id" while the app itself uses the column "_id". I dont know why...but I have to make a check.

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java [hunk 5]
```diff
@@ -303,7 +303,7 @@ public synchronized MultipleChoiceCardDao getMultipleChoiceDao() {
      * Override the finalize in case the helper is not release.
      */
     @Override
-    public void finalize() throws Throwable {
+    protected void finalize() throws Throwable {
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: Why did you change this from public to protected? Does this break backward compatibility?

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java [hunk 6]
```diff
@@ -378,6 +386,36 @@ private void addToMultipleChoiceList(Cursor c, List<MultipleChoiceCard> list) {
     }
 
     public void updateMultipleChoiceId(String newId, String oldId) {
-        db.execSQL("update " + MultipleChoiceContract.MultipleChoiceCardTable.TABLE_NAME +" set id = " + newId + " where id = " + oldId);
+        Cursor res = db.rawQuery("SELECT * FROM " + MultipleChoiceContract.MultipleChoiceCardTable.TABLE_NAME,null);
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: Add a space before null

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java [hunk 7]
```diff
@@ -378,6 +386,36 @@ private void addToMultipleChoiceList(Cursor c, List<MultipleChoiceCard> list) {
...
+        int value = res.getColumnIndex("id");
+        if (value != -1) {
+            db.execSQL("update " + MultipleChoiceContract.MultipleChoiceCardTable.TABLE_NAME +" set id = " + newId + " where id = " + oldId);
+        } else {
+            db.execSQL("update " + MultipleChoiceContract.MultipleChoiceCardTable.TABLE_NAME + " set _id = " + newId + " where _id = " + oldId);
+        }
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: What is the difference between _id and id?

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java [hunk 8]
```diff
@@ -378,6 +386,36 @@ private void addToMultipleChoiceList(Cursor c, List<MultipleChoiceCard> list) {
...
+    }
+
+    public void updateMultipleChoiceCard(MultipleChoiceCard card) {
+        Cursor res = db.rawQuery("SELECT * FROM " + MultipleChoiceContract.MultipleChoiceCardTable.TABLE_NAME,null);
+        int value = res.getColumnIndex("id");
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: Add space before null

File: app/src/main/java/org/liberty/android/fantastischmemo/dao/MultipleChoiceDaoImpl.java [hunk 9]
```diff
@@ -74,10 +74,8 @@ public MultipleChoiceCard getMultipleChoiceCard(long id) {
     public MultipleChoiceCard addMultipleChoiceCard(MultipleChoiceCard mcCard) {
         AnyMemoDBOpenHelper dbHelper = getHelper();
         dbHelper.insertMultipleChoiceCard(mcCard, dbHelper.getWritableDatabase());
-        if (mcCard != null) {
-            return mcCard;
-        }
-        return null;
+
+        return mcCard;
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/dao/MultipleChoiceDaoImpl.java]: Why does an add return a value? Shouldn't it be void?

File: app/src/main/java/org/liberty/android/fantastischmemo/entity/MultipleChoiceCard.java [hunk 10]
```diff
@@ -6,8 +6,12 @@
 
 import org.liberty.android.fantastischmemo.dao.MultipleChoiceDaoImpl;
 
+import java.io.Serializable;
+
 @DatabaseTable(tableName = "multipleChoiceCards", daoClass = MultipleChoiceDaoImpl.class)
-public class MultipleChoiceCard  {
+public class MultipleChoiceCard implements Serializable{
+
+    private static final long serialVersionUID = 2112975910834873118L;
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/entity/MultipleChoiceCard.java]: What's this serialVersionUID used for? Why do the multiple choice cards need to be serializable?

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java [hunk 11]
```diff
@@ -87,7 +91,8 @@ public void onCreate(Bundle savedInstanceState){
         dbOpenHelper = AnyMemoDBOpenHelperManager.getHelper(getApplicationContext(), dbPath);
         multipleChoiceCardDao = dbOpenHelper.getMultipleChoiceDao();
         multipleChoiceCardDao.setHelper(dbOpenHelper);
-        updateViews();
+        setCard(currentMCCard);
+        updateCard(cardToUpdate);
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java]: Why do you have to set the card before updating it? Can't you just update he currentMCCard?

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java [hunk 12]
```diff
@@ -34,6 +33,8 @@
 import org.liberty.android.fantastischmemo.dao.MultipleChoiceCardDao;
 import org.liberty.android.fantastischmemo.entity.MultipleChoiceCard;
 
+import java.util.Objects;
+
 public class CardMCEditor extends BaseActivity{
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java]: Add a space before {

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java [hunk 13]
```diff
@@ -50,11 +51,12 @@
...
+    public static final String EXTRA_DBPATH_MC = "dbpath";
+    public static final String EXTRA_MC = "mcCard";
     String dbPath;
     private AnyMemoDBOpenHelper dbOpenHelper;
-
     private MultipleChoiceCardDao multipleChoiceCardDao;
+    private MultipleChoiceCard cardToUpdate;
 
     @Override
     public void onCreate(Bundle savedInstanceState){
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java]: Add space before {

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java [hunk 14]
```diff
@@ -105,38 +110,66 @@ public void onConfigurationChanged(Configuration newConfig) {
         super.onConfigurationChanged(newConfig);
     }
 
-    public void updateViews(){
+    public void setCard(MultipleChoiceCard card){
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java]: Add space before {

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java [hunk 15]
```diff
@@ -105,38 +110,66 @@ public void onConfigurationChanged(Configuration newConfig) {
...
+            card.setAnswer(option2MCCard);
+        } else if (option3Radio.isChecked()) {
+            card.setAnswer(option3MCCard);
+        } else {
+            card.setAnswer(option4MCCard);
+        }
+    }
+
+    public void updateCard(MultipleChoiceCard card) {
+        if (card != null ){
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java]: Add space between ){

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/OpenActionsMCFragment.java [hunk 16]
```diff
@@ -83,6 +100,35 @@ public void onClick(View v) {
...
+                        .setTitle(getString(R.string.delete_text))
+                        .setMessage(getString(R.string.fb_delete_message))
+                        .setPositiveButton(getString(R.string.delete_text), new DialogInterface.OnClickListener(){
+                            @Override
+                            public void onClick(DialogInterface dialog, int which ){
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/OpenActionsMCFragment.java]: Add space between ){

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/OpenActionsMCFragment.java [hunk 17]
```diff
@@ -83,6 +100,35 @@ public void onClick(View v) {
                 startActivity(myIntent);
                 recentListUtil.addToRecentList(dbPath);
             }
+
+            if(v == deleteItem) {
+                new AlertDialog.Builder(mActivity)
+                        .setTitle(getString(R.string.delete_text))
+                        .setMessage(getString(R.string.fb_delete_message))
+                        .setPositiveButton(getString(R.string.delete_text), new DialogInterface.OnClickListener(){
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/OpenActionsMCFragment.java]: Add space between ){

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/OpenActionsMCFragment.java [hunk 18]
```diff
@@ -83,6 +100,35 @@ public void onClick(View v) {
...
+                                recentListUtil.deleteFromRecentList(dbPath);
+
+                            /* Refresh the list */
+                                mActivity.restartActivity();
+                            }
+                        })
+                        .setNegativeButton(getString(R.string.cancel_text), null)
+                        .create()
+                        .show();
+
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/OpenActionsMCFragment.java]: Remove extra line

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java [hunk 19]
```diff
@@ -411,15 +415,25 @@ public boolean onLongClick(View v) {
                 @Override
                 public void onClick(View v) {
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java]: Whats going on in this method?

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecyclerViewAdapter.java [hunk 20]
```diff
@@ -32,22 +43,63 @@ public MultipleChoiceHolder onCreateViewHolder(ViewGroup parent, int viewType) {
...
+        inflater.inflate(R.menu.mc_card_list_popup_menu, popup.getMenu());
+        popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
+            @Override
+            public boolean onMenuItemClick(MenuItem menuItem) {
+                Intent myIntent = new Intent();
+                switch (menuItem.getItemId()) {
+                    case R.id.delete_mc:
+                        multipleChoiceCardDao.deleteMultipleChoiceCard(card);
+                        mcCards.remove(card);
+                        notifyItemRemoved(position);
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecyclerViewAdapter.java]: Why do you have to delete and then remove the card? Whats the difference?

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecyclerViewAdapter.java [hunk 21]
```diff
@@ -32,22 +43,63 @@ public MultipleChoiceHolder onCreateViewHolder(ViewGroup parent, int viewType) {
...
+                        notifyItemRangeChanged(position, mcCards.size());
+                        break;
+                    case R.id.edit_mc:
+                        myIntent.setClass(mContext, CardMCEditor.class);
+                        myIntent.putExtra(CardMCEditor.EXTRA_DBPATH_MC, dbPath);
+                        myIntent.putExtra(CardMCEditor.EXTRA_MC, card);
+                        mContext.startActivity(myIntent);
+                        break;
+                    default: break;
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecyclerViewAdapter.java]: Put break on the next line

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java (hunk 5 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: I switched it because of the static analysis tool. I will do further testing to see if it broke anything

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java (hunk 7 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: unit tests look for the column "id" while the app itself looks for "_id" so i need to check for both. I dont why it does that. I even tried dropping the tables and re adding them

File: app/src/main/java/org/liberty/android/fantastischmemo/dao/MultipleChoiceDaoImpl.java (hunk 9 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/dao/MultipleChoiceDaoImpl.java]: to be honest I think having it return the card that was added is meaningful. But I can change it to void

File: app/src/main/java/org/liberty/android/fantastischmemo/entity/MultipleChoiceCard.java (hunk 10 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/entity/MultipleChoiceCard.java]: I pass a multiple choice card object in my intent in the recycle view adapter. To pass an object you need to use Serializable. I need to create an id because the static analysis tool said it was required.

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java (hunk 11 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/ui/CardMCEditor.java]: setCard function takes the values that are written in the fields of the ui and updates the card attributes. Ill change it to a more meaningful name. updateCard actually just displays the inital values of the card in the fields of the ui so when you want to edit a card the info is already displayed.

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java (hunk 7 above)
Comment (coreylevine) [reply app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: Ew thats ugly. So much extra code for nothing lol

File: app/src/main/java/org/liberty/android/fantastischmemo/dao/MultipleChoiceDaoImpl.java (hunk 9 above)
Comment (coreylevine) [reply app/src/main/java/org/liberty/android/fantastischmemo/dao/MultipleChoiceDaoImpl.java]: So leave it. If u think its good I'm ok with that. Was just curious.

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecyclerViewAdapter.java (hunk 20 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/ui/RecyclerViewAdapter.java]: delete is removing it from the db. remove is removing it from the current static list. I need to do it for both so it updates the activity automatically without the need to refresh the page

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java (hunk 19 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java]: when you click a deck on the recent list page it's checking if it's a multiple choice deck or a regular deck. And depending on the type it goes to a different a study activity

**Comment by coreylevine** [comment]
@rekandiah Once you make the ASOP changes I will approve the PR.

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java (hunk 7 above)
Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: yeah true. If you have any ideas on how I can minimize it let me know lol

Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: actually i think i know how i can minimize it check it out after
